sdk/go/pulumi/provider.go

```diff
@@ -202,21 +224,316 @@ func constructInputsMap(ctx *Context, inputs map[string]interface{}) (Map, error
 		known := !ci.value.ContainsUnknowns()
 		value, secret, err := unmarshalPropertyValue(ctx, ci.value)
 		if err != nil {
-			return nil, errors.Wrapf(err, "unmarshaling input %s", k)
+			return nil, fmt.Errorf("unmarshaling input %q: %w", k, err)
 		}
 
 		resultType := anyOutputType
 		if ot, ok := concreteTypeToOutputType.Load(reflect.TypeOf(value)); ok {
 			resultType = ot.(reflect.Type)
 		}
 
-		output := ctx.newOutput(resultType, ci.deps...)
+		output := ctx.newOutput(resultType, ci.Dependencies(ctx)...)
 		output.getState().resolve(value, known, secret, nil)
 		result[k] = output
 	}
 	return result, nil
 }
 
+func gatherDeps(v resource.PropertyValue, deps urnSet) {
+	switch {
+	case v.IsSecret():
+		gatherDeps(v.SecretValue().Element, deps)
+	case v.IsComputed():
+		gatherDeps(v.Input().Element, deps)
+	case v.IsOutput():
+		for _, urn := range v.OutputValue().Dependencies {
+			deps.add(URN(urn))
+		}
+		gatherDeps(v.OutputValue().Element, deps)
+	case v.IsResourceReference():
+		deps.add(URN(v.ResourceReferenceValue().URN))
+	case v.IsArray():
+		for _, e := range v.ArrayValue() {
+			gatherDeps(e, deps)
+		}
+	case v.IsObject():
+		for _, e := range v.ObjectValue() {
+			gatherDeps(e, deps)
+		}
+	}
+}
+
+func copyInputTo(ctx *Context, v resource.PropertyValue, dest reflect.Value) error {
```

Comment:
Please help me understand why this is fundamentally different from this functionality in the RPC: 

```
func unmarshalPropertyValue(ctx *Context, v resource.PropertyValue) (interface{}, bool, error) {
```